### ISSUE TITLE
Fail deployments if promotion returns error after deadline

### ISSUE THREAD

**Issue by tgross** [hashicorp/nomad#6740]
Follow-up improvement coming out of https://github.com/hashicorp/nomad/issues/6407#issuecomment-555519437

For each allocation that's part of a deployment, we start up a health hook on the client. Once that hook updates the deployment status of the alloc to healthy (or the deadline passes without doing so), the hook exits. If the allocation fails after it initially reports it's healthy, we're in a state where the deployment thinks all its placements are healthy, but when we promote the deployment to make that a reality, the server's state machine rejects the change because there aren't enough canary allocations that are actually healthy.

This is a pathological case which we're only likely to hit when promotions are made manually or when tasks are slow to start _and_ flappy after start, and getting the behavior to be predictable and understandable to operators in that condition is difficult. There may be some future improvements we can make to the deployments (especially with L7 health checks via Connect on the horizon).

In the meantime, we're going to make a change so that promotions are marked as failed if the promotion hits an error past the promotion deadline.

**Comment by lattwood** [comment]
> slow to start and flappy after start

![Image](https://github.com/user-attachments/assets/332a6d76-2a1f-4288-9202-e290537676f8)

(we are getting burned by the flappy part)

**Comment by lattwood** [comment]
Does the health hook get started when scaling up/down?

**Comment by tgross** [comment]
Yes, the health hooks runs for each allocation on the client independently.
